### PR TITLE
Scene blueprints

### LINKED ISSUE #17: Flag scene import/export
We need some way to edit scenes, but we are too lazy to design level editor ourselves. The solution - use already existing program.
This way we will need to implement only the transition from the editor program to the framework.

Blender allows object to have user defined properties, as well as has support for python-based custom exporters.
That means that it is possible to interpret blender scene objects with specific properties as framework-defined objects and treat them as such during export to custom `.xml` scene descriptor.

## Asset flow

[`.blend` file defining a bunch of semi-empty models (flags) with lists of object-defining properties]

=> *blender-sided python script* =>

[`.xml` scene descriptor]
```XML
<!-- scene descriptor draft -->
<!-- REQUIRES REVIEW AND CONFIRMATION -->

<scene>
    <collider name="_box_house_wall_south">
        <size x="1.0" y="1.0" z="1.0">
        <transform>
            . . .
        </transform>
    </collider>

    <static_mesh name="tree_1024">
        <!-- expands to object group -->
        <file path="models/tree.obj"></file>
        <transform>
            . . .
        </transform>
    </mesh>
    
    <static_mesh name="rock_516">
        <!-- expands to object group -->
        <file path="models/rock.obj"></file>
        <transform>
            . . .
        </transform>
    </mesh>
    
    <spawner name="gazebo_spawner">
        . . .

        <send when="dead" target="fireworks_trigger">
            <set var="state" value="active"></set>
        </send>
    </spawner>

    <box_trigger name="fireworks_trigger">
        <size xyz = . . .><\size>
        <transform>
            . . .
        </transform>
        
        <filter>
            <player></player>
            <entity guid="$shopkeeper_spawner.last_spawned"></entity>
            <entities guids="$gazebo_spawner.fight_participants"></entities>
            
            <effect type="explosion"><\effect>
            <effect type="fire"><\effect>
        </filter>

        <activity state="inactive"></active>
        <send when="enter" target="celebratory_fireworks">
            <trigger event="shoot"></trigger>
        </send>
    </box_trigger>
    
    . . .
</scene>
```

```C++
// Loadable level initialization draft
// REQUIRES REVIEW AND CONFIRMATION

// spawner.h
struct Spawner : public SceneComponent {
    Spawner(const tinyxml2::XMLElement* descriptor);

    . . .
};

// loadable_level.h
struct LoadableLevel : public Scene {
    // Some overridable way of object registration by its descriptor.

   protected:
    friend void read_spawner(LoadableLevel& level, const tinyxml2::XMLElement* descriptor);

    ComponentRegistry<Spawner> spawners_;
};
```

```C++
// loadable_level.cpp

void LoadableLevel::pseudo_constructor() {
    register_reader("spawner", read_spawner);
}

read_spawner(LoadableLevel& level, const tinyxml2::XMLElement* descriptor) {
    Spawner* spawner = level.spawners_.insert(Spawner(descriptor));

    // void Scene::register_component(SceneComponent& component);
    level.register_component(*spawner);
}
```

### PR DESCRIPTION
**Pull request by Sigmarik**
It is now possible to construct scenes in blender. The setup process is a bit tricky, but after it is done scene creation becomes quite pleasant.

## Level assets

`.level.xml` assets store scene component lists that should be present on the level.

```XML
<level>
<component_type_signature name="FirstComponentName">
  <!-- component data -->
</component_type_signature>

<other_cts name="SecondComponentName">
  <!-- component data -->
</other_cts>
</level>
```

Level is a set of components, each of which is defined by the corresponding tag in the level file.

Tag name should be a component's class importer signature, and `name` attribute should contain a unique (for the level) component name, which will be used as its external identifier.

Component data format is defined only by its corresponding importer.

**Level import**

```C++
#include "logics/blueprints/external_level.h"
#include "managers/asset_manager.h"

const ExternalLevel* level = AssetManager::request<ExternalLevel>("my_level.level.xml");

SubcomponentNameMap map = level->build(scene, /* level transform = */ glm::mat4(1.0));

// At this point the scene contains all required components, which can be accessed by name
// using the `map` of the build.
```

**Custom component importer**

```C++
#include "logics/blueprints/component_importer.h"

XML_BASED_IMPORTER(Producer, "custom_signature") {
    // Mandatory parameters can be requested through the `demand` function
    glm::mat4 transform = demand<glm::mat4>(data, "transform", glm::mat4(1.0));

    // Optional parameters can be requested through the `request` function
    int optional_argument = request<int>(data, "optional_parameter", 0);

    // Importer should return a `Producer`, which is a functor which builds the component
    // when the level gets built.

    // `parent_tform` - world space level transform

    // You can either write lambda definition by yourself or use the `PRODUCER` macro
    return new Asset<Producer>([=](const glm::mat4& parent_tform) {
        // Build the component from parameters read in the importer base

        Subcomponent<CustomComponent> component(/* . . . */);
        component->set_transform(parent_tform * transform);
        
        // . . .

        return component;
    });
}
```

Corresponding component tag example:

```XML
<custom_signature name="SomeComponent">
<transform>
<dx x="1.000000" z="0.000000" y="0.000000" w="0.000000"></dx>
<dz x="0.000000" z="1.000000" y="0.000000" w="0.000000"></dz>
<dy x="0.000000" z="-0.000000" y="1.000000" w="0.000000"></dy>
<dw x="0.000000" z="0.500000" y="0.028003" w="1.000000"></dw>
</transform>
<optional_parameter value="1024"></optional_parameter>
</custom_signature>
```

## Formats of library components

**StaticMesh** `static_mesh`
mandatory `mat4` `transform` - component transform
mandatory `string` `source` - mesh source
optional `bool` `collision` - import colliders from the corresponding file

**PointLightComponent** `point_light`
mandatory `mat4` `transform` - light origin transform
mandatory `vec3` `color` - light source color and brightness
optional `float` `radius` - light source radius (0.0 meters y default)
optional `float` `spread` - affected area radius (3.0 meters by default)

**AmbientLightComponent** `ambient_light`
mandatory `vec3` `color` - light color

**StaticBoxCollider** `box_collider`
mandatory `mat4` `transform` - a matrix which transforms centered one meter box into the desired parallelogram .

## Blender as level editor

The pull request contains two extensions for blender, which can help designers with level creation.
One extension implements scene exporter in `.level.xml` format, and the other adds an action to transform the given mesh into framework-recognizable `StaticMesh` descriptor.

### Level exporter

Level exporter recognizes any object with the `type` object property as a framework-viable component, and exports it with the value of the `type` parameter as signature, object name as component name, and all custom object properties as component parameters.

Object transform is written as the `transform` parameter.

custom property list of a static mesh component

It is recommended to use premade component library to avoid manually inputting object mandatory parameters.

**Recognized types**

Point lights marked with the corresponding `type` attribute get their `color` parameter filled as their light color multiplied by the light source power.

The only exporter flag specifies weather the addon should add ambient light component to the scene. In case the ambient light is added the world color is written as its emission color.

**Static mesh workflow tip**

It is useful to have a user-level asset library with objects defined in their corresponding asset files.

For example, a blender file dedicated to the tree model may have a hidden tree-shaped marker set as a static mesh component, which can be placed on a scene with a simple drag-and-drop action from the asset shelf.

## Other improvements

- Updated the `BouncyObject` collision response code so it would not bounce on the floor on lower tickrates,
- Added optional `capture` and `restore` methods to components,
- Fixed component deletion and double registration.

## Restrictions

Component type signature should not be equal to `script`, as this keyword will be used in future versions for script assignment.

More advanced scene-scope components should be defined inside the corresponding `Scene` derivative, as there is no support for these components yet.

Postprocessing volume components may be added soon.

## Possible applications

Level files may not only define complete levels, but parts of one.

Bigger worlds can be defined in a collection of dynamically unloaded levels, and different tiles of a randomly generated maps can be defined in their own separate level descriptors.

## Future work

Level action pipelining seems like a logical next step, along with a framework component library for Blender and support for object linking inside the editor.

## Keywords

Fixes #17

### CODE REVIEW

File: lib/logics/blueprints/component_factory.hpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+   private:
+    std::map<std::string, Producer> instructions_{};
+};
+
+template <class... Ts>
+inline SubcomponentNameMap ComponentFactory<Ts...>::build(Ts&&... args) const {
+    SubcomponentNameMap map{};
+
+    for (auto& [name, producer] : instructions_) {
```
Comment (Nerfiti) [inline lib/logics/blueprints/component_factory.hpp]: const auto&?

File: lib/logics/blueprints/external_level.cpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#include "external_level.h"
+
+#include "logics/scene.h"
+
+SubcomponentNameMap ExternalLevel::build(Scene& scene,
+                                         const glm::mat4& transform) const {
+    SubcomponentNameMap guide = factory_.build(transform);
+
+    for (auto& [name, component] : guide) {
```
Comment (Nerfiti) [inline lib/logics/blueprints/external_level.cpp]: const auto&?

File: lib/logics/subcomponent.hpp [hunk 3]
```diff
@@ -139,6 +173,8 @@ struct WeakSubcomponent final {
         return *this;
     }
 
+    operator bool() const { return static_cast<bool>(lock()); }
```
Comment (Nerfiti) [inline lib/logics/subcomponent.hpp]: Maybe something like that?
```c++
return !expired() && static_cast<bool>(lock());
```

File: lib/logics/blueprints/component_factory.hpp (hunk 1 above)
Comment (Sigmarik) [reply lib/logics/blueprints/component_factory.hpp]: Fixed

File: lib/logics/blueprints/external_level.cpp (hunk 2 above)
Comment (Sigmarik) [reply lib/logics/blueprints/external_level.cpp]: Fixed

File: lib/logics/subcomponent.hpp (hunk 3 above)
Comment (Sigmarik) [reply lib/logics/subcomponent.hpp]: Done

**Review by Nerfiti — APPROVED** [review]

**Pull request merged**